### PR TITLE
Make agent-level code abstraction to unify code structure

### PR DESCRIPTION
**Pull request by kyujin-cho**
## What is this for? 
This PR merges code from feture/kube-integration and master branch into one code base, with code abstraction of whole backend.ai agent structure.

### config.toml
new option '[agent] mode' will be added to`config.toml`. This option is essential. The value should be one of: `docker` or `k8s`. Backend.AI agent will work with mode specified by user.
### package structure
machine-specific codes(used only for docker/k8s) are now splitted into new subpackage:
  - `ai.backend.agent.docker.k8s` for k8s agent
  - `ai.backend.agent.docker.docker` for docker agent
### AbstractKernelRunner
Abstraction for KernelRunner class has been added to `ai.backend.agent.kernel.AbstractKernelRunner`; All machine-specific KernelRunner should inherit this class and implement `start` method.
### AbstractAgentServer
Codes described in `AgentRPCServer`  will now be splited into two parts: `AgentRPCServer` and `AbstractAgentServer`.     
`AgentRPCServer` still remains as a aiozmq RPC Handler class, but all essential implementations (and machine-specific) such as `start_kernel`, `execute`, `start_service` should be implemented machine-specific subpackages, by inheriting `AbstractAgentServer`.

## Remaining duplicates

Almost all duplicated codes for k8s and docker agents has been now unified as a one code; however, there are still some duplicates:
### KernelResourceSpec
`ai.backend.agent.resources.KernelResourceSpec` needs to be defined inside each machine-specific subpackages, since it depends on `known_slot_types` dictionary, which is machine-dependent.

### CODE REVIEW

**Comment by achimnol** [comment]
I cannot review the k8s part combined with restructuring together -- this is tooo big.
Could you split the PR for only restructuring part and the k8s part, so that I could review them separately?

File: src/ai/backend/agent/exception.py [hunk 1]
```diff
@@ -4,3 +4,17 @@ class InitializationError(Exception):
 
 class InsufficientResource(Exception):
     pass
+
+
+class NoKernelError(Exception):
```
Comment (achimnol) [inline src/ai/backend/agent/exception.py]: Could you add a brief docstring about what this exception is for?

File: src/ai/backend/agent/exception.py [hunk 2]
```diff
@@ -4,3 +4,17 @@ class InitializationError(Exception):
...
+    pass
+
+
+class K8sError(Exception):
+    def __init__(self, message):
+        super().__init__(message)
+        self.message = message
+
+
+class ScaleFailedError(Exception):
```
Comment (achimnol) [inline src/ai/backend/agent/exception.py]: Same here.

**Comment by achimnol** [comment]
Also please resolve the conflicts so that I can merge this branch. :wink:

**Comment by kyujin-cho** [comment]
>   * What's the difference to the meta-repository's `install-dev.sh` script?
>   * Could the meta-repository's script updated to work with this generalized agent + k8s branch?)

Wow, I didn't know there already was an installation script! It'll be better to remove the one in this branch and update `install-dev.sh` on meta repository with Kubernetes support.

File: src/ai/backend/agent/exception.py (hunk 1 above)
Comment (kyujin-cho) [reply src/ai/backend/agent/exception.py]: Those exceptions are not used anymore; I'll remove those.

**Pull request merged**